pkg/repository/network.go

```diff
@@ -133,27 +276,427 @@ func (r *networkRepository) Create(ctx context.Context, rq *Validated[*apiv2.Net
+		if err != nil {
+			return nil, errorutil.Convert(err)
+		}
+		newNetwork.DestinationPrefixes = destPrefixes
+	}
+
+	if req.DefaultChildPrefixLength != nil {
+		newDefaultChildPrefixLength, err := metal.ToChildPrefixLength(req.DefaultChildPrefixLength, newNetwork.Prefixes)
+		if err != nil {
+			return nil, errorutil.NewInvalidArgument(err)
+		}
+		newNetwork.DefaultChildPrefixLength = newDefaultChildPrefixLength
+	}
+
+	if req.Force {
+		newNetwork.AdditionalAnnouncableCIDRs = req.AdditionalAnnouncableCidrs
+	}
+
+	r.r.log.Debug("update", "network id", newNetwork.ID, "prefixes to add", prefixesToBeAdded, "prefixes to remove", prefixesToBeRemoved)
+
+	for _, p := range prefixesToBeRemoved {
+		_, err := r.r.ipam.DeletePrefix(ctx, connect.NewRequest(&ipamv1.DeletePrefixRequest{Cidr: p.String(), Namespace: newNetwork.Namespace}))
+		if err != nil {
+			return nil, errorutil.Convert(err)
+		}
+	}
+
+	for _, p := range prefixesToBeAdded {
+		_, err := r.r.ipam.CreatePrefix(ctx, connect.NewRequest(&ipamv1.CreatePrefixRequest{Cidr: p.String(), Namespace: newNetwork.Namespace}))
+		if err != nil {
+			return nil, errorutil.Convert(err)
+		}
+	}
+
+	r.r.log.Debug("updated network", "old", old, "new", newNetwork)
+	newNetwork.SetChanged(old.Changed)
+	err = r.r.ds.Network().Update(ctx, &newNetwork)
+	if err != nil {
+		return nil, err
+	}
+
+	return &newNetwork, nil
 }
-func (r *networkRepository) Find(ctx context.Context, query *apiv2.NetworkServiceListRequest) (*metal.Network, error) {
-	panic("unimplemented")
+
+func (r *networkRepository) Find(ctx context.Context, query *apiv2.NetworkQuery) (*metal.Network, error) {
+	nw, err := r.r.ds.Network().Find(ctx, r.scopedNetworkFilters(queries.NetworkFilter(query))...)
+	if err != nil {
+		return nil, err
+	}
+
+	return nw, nil
 }
-func (r *networkRepository) List(ctx context.Context, query *apiv2.NetworkServiceListRequest) ([]*metal.Network, error) {
-	panic("unimplemented")
+func (r *networkRepository) List(ctx context.Context, query *apiv2.NetworkQuery) ([]*metal.Network, error) {
+	nws, err := r.r.ds.Network().List(ctx, r.scopedNetworkFilters(queries.NetworkFilter(query))...)
+	if err != nil {
+		return nil, err
+	}
+
+	slices.SortFunc(nws, func(a, b *metal.Network) int {
+		return strings.Compare(a.ID, b.ID)
+	})
+
+	return nws, nil
 }
 func (r *networkRepository) ConvertToInternal(msg *apiv2.Network) (*metal.Network, error) {
 	panic("unimplemented")
 }
 func (r *networkRepository) ConvertToProto(e *metal.Network) (*apiv2.Network, error) {
-	panic("unimplemented")
+	var (
+		consumption *apiv2.NetworkConsumption
+		labels      *apiv2.Labels
+		networkType *apiv2.NetworkType
+		natType     *apiv2.NATType
+	)
+
+	if e == nil {
+		return nil, nil
+	}
+
+	if e.Labels != nil {
+		labels = &apiv2.Labels{
+			Labels: e.Labels,
+		}
+	}
+
+	if e.Nat {
+		natType = apiv2.NATType_NAT_TYPE_IPV4_MASQUERADE.Enum()
+	}
+	if e.PrivateSuper {
+		networkType = apiv2.NetworkType_NETWORK_TYPE_SUPER.Enum()
```

Comment:
Check for namespace field missing.

Reply:
namespace cannot be set, its set if super is namespaced with the projectid if the child